File: src/PowerToys/Client/src/power-toy/login-customizer/login-customizer-settings.ts
export const DAYS_OF_WEEK = ["Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"];

export interface LoginCustomizerSettings {
  backgroundImage: string;
  logoImage: string;
  logoImageAlternative: string;
  allowPasswordReset: boolean;
  customCss: string;
  instruction: string;
  /** Login screen greeting overrides, index 0 = Sunday ... 6 = Saturday. */
  greetings: string[];
}

export const DEFAULT_LOGIN_CUSTOMIZER_SETTINGS: LoginCustomizerSettings = {
  backgroundImage: "",
  logoImage: "",
  logoImageAlternative: "",
  allowPasswordReset: true,
  customCss: ":root {\n\n}\n",
  instruction: "",
  greetings: DAYS_OF_WEEK.map(() => ""),
};

// Settings saved before a field existed (or a settings blob that's just missing a key)
// shouldn't leave that field undefined - always fill gaps from the defaults.
export function withDefaults(settings: Partial<LoginCustomizerSettings> | null | undefined): LoginCustomizerSettings {
  return {
    ...DEFAULT_LOGIN_CUSTOMIZER_SETTINGS,
    ...settings,
    greetings: settings?.greetings ?? DEFAULT_LOGIN_CUSTOMIZER_SETTINGS.greetings,
  };
}

File: src/PowerToys/PowerToy/LoginCustomizer/LoginCustomizerSettings.cs
using System.Text.Json;

using PowerToys.Services;

namespace PowerToys.PowerToy.LoginCustomizer
{
    /// <summary>
    ///     The Login Customizer power toy's settings, as saved via the generic settings endpoint.
    /// </summary>
    internal sealed class LoginCustomizerSettings
    {
        public const string PowerToyAlias = "PowerToys.PowerToy.LoginCustomizer";

        public string? BackgroundImage { get; set; }

        public string? LogoImage { get; set; }

        public string? LogoImageAlternative { get; set; }

        public bool AllowPasswordReset { get; set; } = true;

        public string? CustomCss { get; set; }

        public string? Instruction { get; set; }

        /// <summary>
        ///     Login screen greeting overrides, index 0 = Sunday ... 6 = Saturday.
        /// </summary>
        public string?[] Greetings { get; set; } = new string?[7];

        /// <summary>
        ///     Reads the power toy's saved settings, or null if it's disabled or has none saved yet.
        /// </summary>
        public static LoginCustomizerSettings? GetSettings(IPowerToyService powerToyService)
        {
            if (!powerToyService.IsEnabled(PowerToyAlias)) return null;

            var json = powerToyService.GetSettings(PowerToyAlias);
            if (string.IsNullOrWhiteSpace(json)) return null;

            return JsonSerializer.Deserialize<LoginCustomizerSettings>(json, new JsonSerializerOptions
            {
                PropertyNameCaseInsensitive = true,
            });
        }
    }
}
